src/draw/sw/lv_draw_sw_gradient.c

```diff
@@ -55,6 +98,120 @@ static lv_grad_t * allocate_item(const lv_grad_dsc_t * g, int32_t w, int32_t h)
+        146, 147, 148, 149, 150, 150, 151, 152, 153, 154, 155, 155, 156, 157,
+        158, 159, 160, 160, 161, 162, 163, 163, 164, 165, 166, 167, 167, 168,
+        169, 170, 170, 171, 172, 173, 173, 174, 175, 176, 176, 177, 178, 178,
+        179, 180, 181, 181, 182, 183, 183, 184, 185, 185, 186, 187, 187, 188,
+        189, 189, 190, 191, 192, 192, 193, 193, 194, 195, 195, 196, 197, 197,
+        198, 199, 199, 200, 201, 201, 202, 203, 203, 204, 204, 205, 206, 206,
+        207, 208, 208, 209, 209, 210, 211, 211, 212, 212, 213, 214, 214, 215,
+        215, 216, 217, 217, 218, 218, 219, 219, 220, 221, 221, 222, 222, 223,
+        224, 224, 225, 225, 226, 226, 227, 227, 228, 229, 229, 230, 230, 231,
+        231, 232, 232, 233, 234, 234, 235, 235, 236, 236, 237, 237, 238, 238,
+        239, 240, 240, 241, 241, 242, 242, 243, 243, 244, 244, 245, 245, 246,
+        246, 247, 247, 248, 248, 249, 249, 250, 250, 251, 251, 252, 252, 253,
+        253, 254, 254, 255
+    };
+
+    int32_t xn;
+
+    if(x >= 0x10000)
+        if(x >= 0x1000000)
+            if(x >= 0x10000000)
+                if(x >= 0x40000000) {
+                    if(x >= 65535UL * 65535UL)
+                        return 65535;
+                    xn = sqq_table[x >> 24] << 8;
+                }
+                else
+                    xn = sqq_table[x >> 22] << 7;
+            else if(x >= 0x4000000)
+                xn = sqq_table[x >> 20] << 6;
+            else
+                xn = sqq_table[x >> 18] << 5;
+        else {
+            if(x >= 0x100000)
+                if(x >= 0x400000)
+                    xn = sqq_table[x >> 16] << 4;
+                else
+                    xn = sqq_table[x >> 14] << 3;
+            else if(x >= 0x40000)
+                xn = sqq_table[x >> 12] << 2;
+            else
+                xn = sqq_table[x >> 10] << 1;
+
+            goto nr1;
+        }
+    else if(x >= 0x100) {
+        if(x >= 0x1000)
+            if(x >= 0x4000)
+                xn = (sqq_table[x >> 8] >> 0) + 1;
+            else
+                xn = (sqq_table[x >> 6] >> 1) + 1;
+        else if(x >= 0x400)
+            xn = (sqq_table[x >> 4] >> 2) + 1;
+        else
+            xn = (sqq_table[x >> 2] >> 3) + 1;
+
+        goto adj;
+    }
+    else
+        return sqq_table[x] >> 4;
+
+    /* Run two iterations of the standard convergence formula */
+
+    xn = (xn + 1 + x / xn) / 2;
+nr1:
+    xn = (xn + 1 + x / xn) / 2;
+adj:
+
+    if(xn * xn > x)  /* Correct rounding if necessary */
+        xn--;
+
+    return xn;
+}
+
+static inline int32_t sqr32(int32_t x)
+{
+    return x * x;
+}
+
+static inline int32_t extend_w(int32_t w, lv_grad_extend_t extend)
+{
+    if(w < 0)
+        return 0;
+    if(extend == LV_GRAD_EXTEND_PAD) {                  /**< Repeat the same color*/
+        return LV_MIN(w, 255);
+    }
+    if(extend == LV_GRAD_EXTEND_REPEAT) {       /**< Repeat the pattern*/
+        return w & 255;
+    }
+    w &= 511;
```

Comment:
Is this `LV_GRAD_EXTEND_REPEAT`?

Reply:
I don't understand the question.

Reply:
This was a minor performance optimization. First I've implemented this function with a switch(), but the overhead seemed to be significant, compared to this implementation.